libraries/adaptive-expressions/tests/triggerTree.test.js

```diff
@@ -0,0 +1,656 @@
+                );
+                break;
+            case 1:
+                value = Math.random();
+                expression = Expression.makeExpression(
+                    type,
+                    undefined,
+                    Expression.makeExpression(ExpressionType.Accessor, undefined, new Constant(name)),
+                    new Constant(this.adjustDoubleValue(value, type))
+                );
+                break;
+        }
+        const expressionInfo = {
+            expression,
+            bindings: {},
+            quantifiers: []
+        };
+        expressionInfo.bindings[name] = { type, value };
+        return expressionInfo;
+    }
+
+    generateHasValueComparison(name) {
+        let expression;
+        let value;
+        const selection = Math.floor(Math.random() * 3);
+        switch (selection) {
+            case 0:
+                expression = Expression.makeExpression(
+                    ExpressionType.Exists,
+                    undefined,
+                    Expression.makeExpression(ExpressionType.Accessor, undefined, new Constant(name))
+                );
+                value = Math.floor(Math.random() * 2147483647);
+                break;
+            case 1:
+                expression = Expression.makeExpression(
+                    ExpressionType.Exists,
+                    undefined,
+                    Expression.makeExpression(ExpressionType.Accessor, undefined, new Constant(name))
+                );
+                value = Math.random();
+                break;
+            case 2:
+                expression = Expression.makeExpression(
+                    ExpressionType.NotEqual,
+                    undefined,
+                    Expression.makeExpression(ExpressionType.Accessor, undefined, new Constant(name)),
+                    new Constant(null)
+                );
+                value = this.randomString(5);
+                break;
+        }
+
+        const expressionInfo = {
+            expression,
+            bindings: {},
+            quantifiers: []
+        };
+        expressionInfo.bindings[name] = {
+            value, type: ExpressionType.Not
+        };
+        return expressionInfo;
+    }
+
+    mergeBindings(expressions) {
+        const bindings = {};
+        for (const info of expressions) {
+            for (const key in info.bindings) {
+                bindings[key] = info.bindings[key];
+            }
+        }
+        return bindings;
+    }
+
+    randomString(length) {
+        let result = '';
+        for (let i = 0; i < length; i++) {
+            const chars = 'abcdefghijklmnopqrstuvwxyz';
+            const index = Math.floor(Math.random() * 26);
+            result += chars[index];
+        }
+        return result;
+    }
+
+    randomChoice(choices) {
+        const selection = Math.floor(Math.random() * choices.length);
+        return choices[selection];
+    }
+
+    randomWeighted(weights) {
+        let totalWeight = 0;
```

Comment:
Up to you here, but `Array.prototype.reduce` is handy for this.
```typescript
const totalWeight = weights.reduce((sum, weight) => sum + weight, 0);
```

Reply:
Thanks for reminding this.